service/history/workflowTaskHandler.go

```diff
@@ -361,33 +360,25 @@ func (handler *workflowTaskHandlerImpl) handleCommandCompleteWorkflow(
 		return nil
 	}
 
-	// check if this is a cron workflow
-	cronBackoff, err := handler.mutableState.GetCronBackoffDuration()
-	if err != nil {
-		handler.stopProcessing = true
-		return err
-	}
-	if cronBackoff == backoff.NoBackoff {
-		// not cron, so complete this workflow execution
-		if _, err := handler.mutableState.AddCompletedWorkflowEvent(handler.workflowTaskCompletedID, attr); err != nil {
+	// Check if this workflow has a cron schedule
+	if cronBackoff := handler.mutableState.GetCronBackoffDuration(); cronBackoff != backoff.NoBackoff {
+		_, err = handler.mutableState.AddCompletedWorkflowEvent(
+			handler.workflowTaskCompletedID,
+			attr)
+		if err != nil {
 			return serviceerror.NewInternal("Unable to add complete workflow event.")
 		}
-		return nil
+		return handler.handleCron(cronBackoff, attr.GetResult(), nil)
 	}
 
-	// this is a cron workflow
-	startEvent, err := handler.mutableState.GetStartEvent()
+	// Not cron, so complete this workflow execution
+	_, err = handler.mutableState.AddCompletedWorkflowEvent(
+		handler.workflowTaskCompletedID,
+		attr)
 	if err != nil {
-		return err
+		return serviceerror.NewInternal("Unable to add complete workflow event.")
 	}
```

Comment:
maybe 
```go
cronBackoff := handler.mutableState.GetCronBackoffDuration()
_, err = handler.mutableState.AddCompletedWorkflowEvent(
	handler.workflowTaskCompletedID,
	attr,
)
if err != nil {
	return serviceerror.NewInternal("Unable to add complete workflow event.")
}
if cronBackoff == backoff.NoBackoff {
	return nil
}
return handler.handleCron(cronBackoff, attr.GetResult(), nil)
```

Reply:
yeah, I tried like eight different ways to rearrange this... it's a balance between clearly splitting the cron (or retry) path, and reducing code duplication. it helped me understand things to go more on the side of cleanly splitting at firt, but I should probably merge them back more to reduce duplication. the failure case is messier, though.